src/fastfield/alive_bitset.rs

```diff
@@ -0,0 +1,200 @@
+use crate::directory::FileSlice;
+use crate::directory::OwnedBytes;
+use crate::space_usage::ByteCount;
+use crate::DocId;
+use common::BitSet;
+use std::io;
+use std::io::Write;
+
+/// Write a alive `BitSet`
+///
+/// where `alive_bitset` is the set of alive `DocId`.
+/// Warning: this function does not call terminate. The caller is in charge of
+/// closing the writer properly.
+pub fn write_alive_bitset<T: Write>(alive_bitset: &BitSet, writer: &mut T) -> io::Result<()> {
+    alive_bitset.serialize(writer)?;
+    Ok(())
+}
+
+/// Set of alive `DocId`s.
+#[derive(Clone)]
+pub struct AliveBitSet {
+    data: OwnedBytes,
+    num_deleted: usize,
+}
+
+impl AliveBitSet {
+    #[cfg(test)]
+    pub(crate) fn for_test(not_alive_docs: &[DocId], max_doc: u32) -> AliveBitSet {
+        use crate::directory::{Directory, RamDirectory, TerminatingWrite};
+        use std::path::Path;
+        assert!(not_alive_docs.iter().all(|&doc| doc < max_doc));
+        let mut bitset = BitSet::with_max_value_and_filled(max_doc);
+        for &doc in not_alive_docs {
+            bitset.remove(doc);
+        }
+        let directory = RamDirectory::create();
+        let path = Path::new("dummydeletebitset");
+        let mut wrt = directory.open_write(path).unwrap();
+        write_alive_bitset(&bitset, &mut wrt).unwrap();
+        wrt.terminate().unwrap();
+        let file = directory.open_read(path).unwrap();
+        Self::open(file).unwrap()
+    }
+
+    /// Opens a delete bitset given its file.
+    pub fn open(file: FileSlice) -> crate::Result<AliveBitSet> {
+        let bytes = file.read_bytes()?;
+        let num_deleted = BitSet::count_unset_from_bytes(bytes.as_slice());
+
+        Ok(AliveBitSet {
+            data: bytes,
+            num_deleted,
+        })
+    }
+
+    /// Returns true iff the document is still "alive". In other words, if it has not been deleted.
+    #[inline]
+    pub fn is_alive(&self, doc: DocId) -> bool {
+        !self.is_deleted(doc)
+    }
+
+    /// Returns true iff the document has been marked as deleted.
+    #[inline]
+    pub fn is_deleted(&self, doc: DocId) -> bool {
+        let data = self.data.as_slice();
+        !BitSet::contains_from_bytes(doc, data)
+    }
+
+    /// Iterate over the positions of the set elements
+    #[inline]
+    pub fn iter_unset(&self) -> impl Iterator<Item = u32> + '_ {
```

Comment:
s/u32/DocId

Reply:
iter_deleted

Reply:
and actually we don't care about iter_deleted,(and iter_unset is erroneous)...

What we want is iter_alive